src/tensors/diagtensor.jl

```diff
@@ -0,0 +1,113 @@
+# DiagonalTensorMap
+#==========================================================#
+struct DiagonalTensorMap{T,S<:IndexSpace,A<:DenseVector{T}} <: AbstractTensorMap{T,S,1,1}
+    data::A
+    domain::S # equals codomain
+
+    # uninitialized constructors
+    function DiagonalTensorMap{T,S,A}(::UndefInitializer,
+                                      dom::S) where {T,S<:IndexSpace,A<:DenseVector{T}}
+        data = A(undef, reduceddim(dom))
+        return DiagonalTensorMap{T,S,A}(data, dom)
+    end
+    # constructors from data
+    function DiagonalTensorMap{T,S,A}(data::A,
+                                      dom::S) where {T,S<:IndexSpace,A<:DenseVector{T}}
+        T ⊆ field(S) || @warn("scalartype(data) = $T ⊈ $(field(S)))", maxlog = 1)
+        return DiagonalTensorMap{T,S,A}(data, dom)
+    end
+end
+reduceddim(V::IndexSpace) = sum(c -> dim(V, c), sectors(V); init=0)
+
+# Basic methods for characterising a tensor:
+#--------------------------------------------
+space(t::DiagonalTensorMap) = t.domain ← t.domain
+
+"""
+    storagetype(::Union{T,Type{T}}) where {T<:TensorMap} -> Type{A<:DenseVector}
+
+Return the type of the storage `A` of the tensor map.
+"""
+storagetype(::Type{<:DiagonalTensorMap{T,S,A}}) where {T,S,A<:DenseVector{T}} = A
+
+# DiagonalTensorMap constructors
+#--------------------------------
+# undef constructors
+"""
+    DiagonalTensorMap{T}(undef, domain::S) where {T,S<:IndexSpace}
+    # expert mode: select storage type `A`
+    DiagonalTensorMap{T,S,A}(undef, domain::S) where {T,S<:IndexSpace,A<:DenseVector{T}}
+
+Construct a `DiagonalTensorMap` with uninitialized data.
+"""
+function DiagonalTensorMap{T}(::UndefInitializer, V::S) where {T,S<:IndexSpace}
+    return DiagonalTensorMap{T,S,Vector{T}}(undef, V)
+end
+
+function DiagonalTensorMap{T}(data::A, V::S) where {T,S<:IndexSpace,A<:DenseVector{T}}
+    length(data) == reduceddim(V) ||
+        throw(DimensionMismatch("length(data) = $(length(data)) is not compatible with the space $V"))
+    return DiagonalTensorMap{T,S,A}(data, V)
+end
+
+function DiagonalTensorMap(data::DenseVector{T}, V::IndexSpace) where {T}
+    return DiagonalTensorMap{T}(data, V)
+end
+
+# TODO: more constructors needed?
+
+# Special case adjoint:
+#-----------------------
+Base.adjoint(t::DiagonalTensorMap{<:Real}) = t
+Base.adjoint(t::DiagonalTensorMap{<:Complex}) = DiagonalTensorMap(conj(t.data), t.domain)
```

Comment:
Is this a temporary solution, or would you prefer to not have to deal with `AdjointTensorMap{DiagonalTensorMap}`? In principle, in the current approach, this would work, but I agree that handling the dispatch is annoying

Reply:
I tried to mimick `LinearAlgebra.Diagonal`. It seems that the adjoint of a complex `Diagonal` object becomes:
```julia
Diagonal{ComplexF64, Base.ReshapedArray{ComplexF64, 1, Adjoint{ComplexF64, Vector{ComplexF64}}, Tuple{}}}
```
So it is not simply using the `Adjoint` wrapper, but makes sure the outer type is `Diagonal` again. I also think it is useful to make the adjoint be a `DiagonalTensorMap` instance again, so this was the easiest solution.